applications/serial_lte_modem/src/gpio/slm_at_gpio.c

```diff
@@ -0,0 +1,312 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <logging/log.h>
+#include <zephyr.h>
+#include <stdio.h>
+#include "slm_util.h"
+#include "slm_at_gpio.h"
+
+LOG_MODULE_REGISTER(slm_gpio, CONFIG_SLM_LOG_LEVEL);
+
+/* global functions defined in different resources */
+void rsp_send(const uint8_t *str, size_t len);
+
+/* global variable defined in different resources */
+extern struct at_param_list at_param_list;
+extern char rsp_buf[CONFIG_SLM_SOCKET_RX_MAX * 2];
+
+const struct device *gpio_dev;
+static sys_slist_t slm_gpios = SYS_SLIST_STATIC_INIT(&slm_gpios);
+
+/* global variable defined in different files */
+extern struct k_work_q slm_work_q;
+
+struct slm_gpio_pin_t {
+	sys_snode_t node;
+	gpio_pin_t pin;
+	uint16_t op;
+};
+
+gpio_flags_t convert_flags(uint16_t op)
+{
+	gpio_flags_t gpio_flags = UINT32_MAX;
+
+	switch (op) {
+	case SLM_GPIOC_OP_DISABLE:
+		gpio_flags = GPIO_DISCONNECTED;
+		break;
+	case SLM_GPIOC_OP_OUT:
+		gpio_flags = GPIO_OUTPUT;
+		break;
+	case SLM_GPIOC_OP_IN_PU:
+		gpio_flags = GPIO_INPUT | GPIO_PULL_UP;
+		break;
+	case SLM_GPIOC_OP_IN_PD:
+		gpio_flags = GPIO_INPUT | GPIO_PULL_DOWN;
+		break;
+	default:
+		LOG_ERR("Fail to convert gpio flag");
+		break;
+	}
+
+	return gpio_flags;
+}
+
+int do_gpio_pin_configure_set(uint16_t op, gpio_pin_t pin)
+{
+	int err = 0;
+	gpio_flags_t gpio_flags = 0;
+	struct slm_gpio_pin_t *slm_gpio_pin = NULL, *cur = NULL, *next = NULL;
+	gpio_port_pins_t pin_mask = 0;
+
+	LOG_INF("op:%hu pin:%hu", op, pin);
+
+	/* Verify pin correctness */
+	if (pin > MAX_GPIO_PIN) {
+		LOG_ERR("Incorrect <pin>: %d", pin);
+		return -EINVAL;
+	}
+
+	/* Convert SLM GPIO flag to zephyr gpio pin configuration flag */
+	gpio_flags = convert_flags(op);
+	if (gpio_flags == UINT32_MAX) {
+		LOG_ERR("Fail to configure pin.");
+		return -EINVAL;
+	}
+
+	/* Trace gpio list */
+	if (sys_slist_peek_head(&slm_gpios) != NULL) {
+		SYS_SLIST_FOR_EACH_CONTAINER_SAFE(&slm_gpios, cur,
+					     next, node) {
+			if (cur->pin == pin) {
+				slm_gpio_pin = cur;
+			}
+		}
+	}
+
+	/* Add GPIO node if node does not exist */
+	if (slm_gpio_pin == NULL) {
+		slm_gpio_pin = (struct slm_gpio_pin_t *)k_malloc(sizeof(struct slm_gpio_pin_t));
+		if (slm_gpio_pin == NULL) {
+			return -ENOBUFS;
+		}
+		memset(slm_gpio_pin, 0, sizeof(struct slm_gpio_pin_t));
+		sys_slist_append(&slm_gpios, &slm_gpio_pin->node);
+	}
+
+	err = gpio_pin_configure(gpio_dev, pin, gpio_flags);
+	if (err) {
+		LOG_ERR("GPIO_0 config error: %d", err);
+		//TODO: free and remove node
+	}
+
+	slm_gpio_pin->pin = pin;
+	slm_gpio_pin->op = op;
+
+	if (op == SLM_GPIOC_OP_DISABLE) {
+		/* Disable interrupt */
+		err = gpio_pin_interrupt_configure(gpio_dev, pin, GPIO_INT_DISABLE);
+		if (err) {
+			LOG_ERR("Interface pin interrupt config error: %d", err);
+			return err;
+		}
+		/* Remove callback */
+		pin_mask &= ~BIT(pin);
```

Comment:
Once you disable GPIO, shouldn't you free the allocated memory as well? Or is it such a design, that once you allocate a memory for a GPIO, it stays allocated till reboot? If so, perhaps it'd be better to pre-allocate the memory statically, and readuce the heap size?

Reply:
Remove slm_gpio_pin in sys_lists and free slm_gpio_pin.